Initial implementation

forsyth2 opened the pull request:
Initial implementation for `zppy-interfaces`. Resolves discussion at https://github.com/E3SM-Project/zppy/discussions/641. Implements https://github.com/E3SM-Project/zppy/issues/398.

This pull request should be reviewed in conjunction with https://github.com/E3SM-Project/zppy/pull/642.

forsyth2 commented on zppy-interfaces/__main__.py:
```diff
@@ -0,0 +1,28 @@
+    # exclude the rest of the args too, or validation will fail
+    args: argparse.Namespace = parser.parse_args(sys.argv[1:2])
+
+    if args.interface == "global_time_series":
+        global_time_series.global_time_series()
```
I'm following `zstash`'s design here: have an arg parser for the package as a whole (e.g., https://github.com/E3SM-Project/zstash/blob/main/zstash/main.py#L34) and for the individual utilities [commands in `zstash` or `interfaces` in this package] (e.g., https://github.com/E3SM-Project/zstash/blob/main/zstash/create.py#L103)

The result is that calling the global time series code would be done via:
```
zppy-interfaces global-time-series <args>
```

xylar replied:
@forsyth2, this seems like a bit of a clunky interface to me.  Since the tools don't have much to unify them, I would just given them individual entry points.  You could give them a unique prefix like `zi-global-time-series` if you like.

`zstash` is a coherent package with clear subcommands like `git` or `conda` so it makes sense to use this kind of an approach there but I think zppy-interfaces doesn't clearly benefit in the same way.

xylar replied:
This is obviously just my opinion.  You can stick with the `zppy-intefaces` main command with subcommands if you like.

forsyth2 replied:
@xylar If we do completely separate commands, would it make more sense to have distinct dependency requirement lists too? (As opposed to a single `dev.yml` for all of `zppy-interfaces`). I'm not sure if that was something we had decided on.

forsyth2 replied:
I could see several utilities being useful to multiple interfaces -- for example, the `OutputViewer` class for Global Time Series that https://github.com/E3SM-Project/zppy/pull/616 will introduce will also be useful for adding Viewers to the PMP task in https://github.com/E3SM-Project/zppy/pull/640. It follows that it may be useful to have a single dev environment for all of `zppy-interfaces`.

xylar replied:
> @xylar If we do completely separate commands, would it make more sense to have distinct dependency requirement lists too? (As opposed to a single `dev.yml` for all of `zppy-interfaces`). I'm not sure if that was something we had decided on.


No, I would just have a single set of dependencies that covers the full package (all commands and any public functions, etc.).

xylar replied:
> I could see several utilities being useful to multiple interfaces -- for example, the `OutputViewer` class for Global Time Series that https://github.com/E3SM-Project/zppy/pull/616 will introduce will also be useful for adding Viewers to the PMP task in https://github.com/E3SM-Project/zppy/pull/640. It follows that it may be useful to have a single dev environment for all of `zppy-interfaces`.

Yes, and there should be a place for a common framework that includes things like this.

forsyth2 commented on zppy-interfaces/global-time-series/global_time_series.py:
```diff
@@ -0,0 +1,84 @@
+
+from typing import Dict
+
+def main():
```
This function actually replaces much of the code in the `global_time_series.bash` template, bringing us closer to the goal of the bash templates being purely for launching jobs (rather than doing setup work).

xylar replied:
Related to may comment above, I would just make this function the entry point for a `zi-global-time-series` command.

forsyth2 commented on zppy-interfaces/global-time-series/global_time_series.py:
```diff
@@ -0,0 +1,84 @@
+
+def _get_args() -> argparse.Namespace:
+    # Parser
+    parser: argparse.ArgumentParser = argparse.ArgumentParser(
```
`global_time_series` requires a large number of parameters. A `cfg` like `e3sm_diags` has may prove more useful than a bunch of command line arguments. I.e., in https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/e3sm_diags.bash:
```
command="srun -n 1 python -u e3sm.py -d e3sm_diags.cfg"
```

xylar replied:
There can be benefits in supporting both approaches.  All flags could be optional but they would override config options if provided.  Similarly, all user (or zppy-provided) config options could be optional but they would override defaults.  This approach would be attractive for standalone runs but maybe could become confusing through zppy so maybe for that reason it's safer to use one or the other approach -- config options or command-line flags.

forsyth2 commented on zppy-interfaces/global-time-series/global_time_series.py:
```diff
@@ -0,0 +1,84 @@
+        "plots_ocn": args.plots_ocn,
+        "regions": args.regions,
+    }
+    coupled_global.coupled_global(parameters)
```
https://github.com/E3SM-Project/zppy/pull/616 itself refactors the `coupled_global.py` code, in addition to adding a number of unit tests. So, that pull request's code will just be moved over to this repo after this pull request merges.

forsyth2 reviewed (commented):
@xylar @chengzhuzhang I have more work to do, but I think I have the main idea of the refactor captured in this pull request and the corresponding https://github.com/E3SM-Project/zppy/pull/642. 

I should stress I haven't run this code yet. I still need to get the environments sorted out to be able to call `zppy-interfaces` from `zppy`. That is, while I think the main idea is captured here, there is still polishing/testing work left. However, I wanted to tag you here to allow for any further design input early on.

Specifically, my remaining action items:

- [ ] Create a `conda` directory and setup the yaml/toml files necessary to produce a dev environment.
- [x] Use `zppy` to generate stand-alone `ts` and `mpas_analysis` output.
- [x] Run `global-time-series` interface in the dev environment on that output.
- [ ] Run the `global_time_series` task in zppy (i.e., make sure `zppy` can correctly call `zppy-interfaces`).
- [ ] Create a `tests/` directory and add some integration tests. `zppy` should simply test that it can invoke `zppy-interfaces global-time-series`. `zppy-interfaces` however should test that Global Time Series plots actually generate correctly. As for unit tests, https://github.com/E3SM-Project/zppy/pull/616 will do that once it's ported to this repo.
- [ ] Add pre-commit checks
- [ ] Add documentation for this repo/package.

xylar commented on zppy-interfaces/global-time-series/readTS.py:
```diff
@@ -0,0 +1,78 @@
+from typing import Optional, Tuple
```
Please rename this file `read_ts.py` or `read_t_s.py`.

xylar replied:
Actually, the class is `TS`, so the file should be `ts.py`.

xylar commented on zppy-interfaces/global-time-series/ocean_month.py:
```diff
@@ -0,0 +1,132 @@
+from netCDF4 import Dataset, chartostring, date2num
+
+def ocean_month(path_in: str, case_dir: str, start_yr: int, end_yr: int, ts_num_years: int):
+    path_out = "{}/post/ocn/glb/ts/monthly/{}yr".format(case_dir, ts_num_years)
```
Consider switching all format and `%` strings to f-strings as part of this move.  Consider linting the code in other ways.  It's a wonderful chance to clean things up!

Consider adding docstrings to all public functions, methods and classes.

xylar commented on zppy-interfaces/global-time-series/readTS.py:
```diff
@@ -0,0 +1,78 @@
+        # Is this a derived variable?
+        if var == "RESTOM":
+
+            FSNT, _ = self.globalAnnual("FSNT")
```
Consider enforcing that all variables are lowercase with underscores.  At the very least, consider avoiding variables that are all caps.

forsyth2 commented:
Thanks @xylar!

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,40 @@
+# Conda development environment for testing local source code changes to `zppy-interfaces` before merging them to production (`master` branch).
```
```suggestion
# Conda development environment for testing local source code changes to zppy-interfaces before merging them to production (main branch).
```
You can keep the back ticks if you want but this isn't markdown so it's not conventional.

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,40 @@
+  - python >=3.9
+  - pip
+  - matplotlib-base
+  - mpas_tools
```
```suggestion
  - mpas_tools >=0.21.0
```
This is the most recent version that changed CIME constants.

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,40 @@
+  # =================
+  - python >=3.9
+  - pip
+  - matplotlib-base
```
Try to find out what constraint is needed here.  Presumably, not all versions will be okay.

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,40 @@
+  - netcdf4
+  - numpy >=2.0,<3.0
+  - xarray >=2023.02.0
+  - xcdat
```
Again, probably there's a constraint here.

xylar reviewed (commented):
A few comments.

forsyth2 commented on zppy_interfaces/global_time_series/global_time_series.py:
```diff
@@ -0,0 +1,179 @@
+###############################################################################
+
+
+def main(parameters=None):
```
Naming and organization:

`mypy` gave me `zppy-interfaces is not a valid Python package name`. I changed the directory names to use underscores: `zppy_interfaces` and `global_time_series`, and then all the pre-commit checks passed.

3 further questions:
1. Should this file be named `zppy_interfaces/global_time_series/global_time_series.py` [current] or `zppy_interfaces/global_time_series/__main__.py`?
2. Is this function stylistically correctly named `main`?
3. Should we organize interfaces together with their tests or with each other? I.e. which of these:
  - `zppy_interfaces/global_time_series` and `tests/global_time_series` [current implementation, aside from dashes in name currently]
  - `global_time_series/src` and `global_time_series/tests`

xylar replied:
Correct, python cannot have dashes in any of its file names.

xylar replied:
2. yes, that's correct.

xylar replied:
3. tests should be outside of the `zppy_interfaces` package so your first suggestion is far better.

xylar replied:
My mistake on 1.  I took a closer look and everything you have does belong in `__main__.py` so leave it as you have it.

forsyth2 commented on zppy_interfaces/global_time_series/global_time_series.py:
```diff
@@ -0,0 +1,179 @@
+    coupled_global.coupled_global(parameters)
+
+
+# TODO: replace command line arguments with _get_cfg_parameters, like https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/parser/core_parser.py#L809
```
I do think a `cfg` would be cleaner, but I want to get this all working properly first before spending time on that.

forsyth2 reviewed (commented):
Notes on latest changes

forsyth2 commented on zppy_interfaces/global_time_series/coupled_global.py:
```diff
@@ -0,0 +1,1036 @@
+        set_var(
+            exp, "land", requested_variables.vars_land, valid_vars, invalid_vars, rgn
+        )
+        # TODO: Fails here when atmosphere_only="false"
+        set_var(
+            exp, "ocean", requested_variables.vars_ocn, valid_vars, invalid_vars, rgn
+        )
```
(I'm waiting for nodes to run `zppy`'s `main` branch with this configuration to see if something broke there or if the problem is specific to this PR).

xylar replied:
Can you provide the full stack trace for the error?  It presumably comes from a call to `xarray.concat()` and it may be that you need to explicitly specify `dim='Time'` or something like that.  You may also need to have other arguments.  That's where I'd need to see the specifics of the call.  Ocean input files don't have a CF-compliant time coordinate, which causes issues.  (Omega will fix this...)

forsyth2 replied:
It appears this was related to the `xcdat` issue. Per https://github.com/E3SM-Project/zppy-interfaces/pull/1#issuecomment-2492057597, I'm able to generate all plots -- ocean plots included.

xylar commented on conda/dev.yml:
```diff
@@ -8,12 +8,12 @@ dependencies:
   - pip
-  - matplotlib-base
-  - mpas_tools
+  - matplotlib-base # TODO: Find constraint
```
If you know nothing else, I would use the versions pinned in E3SM-Unified as the lower bounds and not have an upper bound.  However, with `matplotlib`, it might be best to have an upper bound for safety, since there are frequent, breaking changes:
https://github.com/MPAS-Dev/geometric_features/blob/93f71ba01d0722de7bfb1b9ec3e9129b859b8963/feature_creation_scripts/ismip6_landice_basins/basin_masks_to_geojson_regions.py#L116
```suggestion
  - matplotlib-base >=3.8.2,<3.10
```
The current version is 3.9.2 so this would be allowed, but 3.10 is under development and we could test it out before officially loosening the constraints to allow it in a future release.

xylar commented on conda/dev.yml:
```diff
@@ -8,12 +8,12 @@ dependencies:
   - numpy >=2.0,<3.0
   - xarray >=2023.02.0
-  - xcdat
+  - xcdat # TODO: Find constraint
```
```suggestion
  - xcdat >=0.7.0,<1.0
```
Again, the lower bound is the version pinned in the last E3SM-Unified, the upper bound is the next major version, assuming that would likely have breaking changes.

@tomvothecoder, should this be tighter, e.g. 0.8.0? Often, minor versions have breaking changes for packages before the 1.0.0 release.  Is that true of xcdat?

tomvothecoder replied:
xCDAT is becoming more stable with each new release. If there are breaking changes, we usually have deprecation warnings. However, I do think it is a good idea to use `xcdat >=0.7.0,<1.0` until a 1.0 is officially released.

xylar commented on setup.cfg:
```diff
@@ -0,0 +1,54 @@
+[flake8]
```
Most of this should move to `pyproject.toml`.  See https://github.com/E3SM-Project/mache/blob/main/pyproject.toml

Frustratingly flake8 refuses to support this format so it has to move to a `.flake8` file:
https://github.com/E3SM-Project/mache/blob/main/.flake8

tomvothecoder replied:
I moved to `ruff` for code formatting and linting, which removed the need for the `setup.cfg` file just for flake8.

xylar commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+# This config file is required for Black.
+# Unforunately, black does not support setup.cfg (refer to link).
+# https://github.com/psf/black/issues/683#issuecomment-490236406
```
This can be removed, because we don't want to use a setup.cfg anymore anyway.

xylar commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+# This config file is required for Black.
+# Unforunately, black does not support setup.cfg (refer to link).
+# https://github.com/psf/black/issues/683#issuecomment-490236406
+
```
Someone (@altheaden or @forsyth2) will need to add these sections:
https://github.com/E3SM-Project/mache/blob/main/pyproject.toml#L1-L50
https://github.com/E3SM-Project/mache/blob/main/pyproject.toml#L67-L83

xylar replied:
The dependencies (build or runtime) listed here must be available on PyPI (installable from `pip`) so that won't include `nco`, `mpas_tools` and perhaps others.  The redundancy between these dependencies and `def.yml` is a little frustrating but this version allows the use of the `pip check` command to determine if the packages in a conda environment are, in fact, consistent with the constraints.  It can sometimes catch problems with dependencies at that have not correctly been updated (e.g. on conda-forge).

xylar replied:
Excellent, looks great!

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,40 @@
+  # Base
+  # =================
+  - python >=3.9
+  - pip
```
This should be moved do a `# Build` section, since it's not needed at runtime.

xylar commented on pyproject.toml:
```diff
@@ -0,0 +1,82 @@
+]
+
+dependencies = [
+    "importlib_resources",
+    "jinja2",
+    "lxml",
+    "pyyaml",
+    "progressbar2",
```
These need to be the same as the base in dev.yml

xylar replied:
Not mpas_tools

altheaden replied:
This is all the packages but they can be split into different sections as they are in the dev.yml file
```suggestion
    "python >=3.9",
    "pip",
    "matplotlib-base >=3.8.2,<3.10",
    "netcdf4",
    "numpy >=2.0,<3.0",
    "xarray >=2023.02.0",
    "xcdat >=0.7.0,<1.0"
    
    "pytest",
    "pytest-cov",
    
    "sphinx",
    "sphinx_rtd_theme",
    "sphinx-multiversion",
    
    "black=23.9.1",
    "flake8=6.1.0",
    "flake8-isort=6.1.0",
    "isort=5.12.0",
    "mypy=1.5.1",
    "pre-commit >=3.0.0",
    "types-PyYAML >=6.0.0",
    
    "tbump=6.9.0",
    "ipykernel",
```

Breaking them up would look like this (added down here because it is not in the range of the suggestion comment):
```toml
dependencies = [
    "python >=3.9",
    "pip",
    "matplotlib-base >=3.8.2,<3.10",
    "netcdf4",
    "numpy >=2.0,<3.0",
    "xarray >=2023.02.0",
    "xcdat >=0.7.0,<1.0",
]

[project.optional-dependencies]

testing = [
    "pytest",
    "pytest-cov",
]

docs = [
    "sphinx",
    "sphinx_rtd_theme",
    "sphinx-multiversion",
]

qa = [
    "black=23.9.1",
    "flake8=6.1.0",
    "flake8-isort=6.1.0",
    "isort=5.12.0",
    "mypy=1.5.1",
    "pre-commit >=3.0.0",
    "types-PyYAML >=6.0.0",
]

dev = [
    "tbump=6.9.0",
    "ipykernel",
]
```

forsyth2 replied:
Hmm, changing this seems to break my `pip install .`: `configuration error: `project.optional-dependencies.qa[0]` must be pep508`

<details>

<summary> Stack trace </summary>

```
Processing /gpfs/fs1/home/ac.forsyth2/ez/zppy-interfaces
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [53 lines of output]
      configuration error: `project.optional-dependencies.qa[0]` must be pep508
      DESCRIPTION:
          Project dependency specification according to PEP 508
      
      GIVEN VALUE:
          "black=23.9.1"
      
      OFFENDING RULE: 'format'
      
      DEFINITION:
          {
              "$id": "#/definitions/dependency",
              "title": "Dependency",
              "type": "string",
              "format": "pep508"
          }
      
      For more details about `format` see
      https://validate-pyproject.readthedocs.io/en/latest/api/validate_pyproject.formats.html
      
      Traceback (most recent call last):
        File "/home/ac.forsyth2/miniconda3/envs/zppy-interfaces-dev-20241104/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/ac.forsyth2/miniconda3/envs/zppy-interfaces-dev-20241104/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/home/ac.forsyth2/miniconda3/envs/zppy-interfaces-dev-20241104/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 333, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 303, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 319, in run_setup
          exec(code, locals())
        File "<string>", line 1, in <module>
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 117, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 157, in setup
          dist.parse_config_files()
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/dist.py", line 655, in parse_config_files
          pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/config/pyprojecttoml.py", line 71, in apply_configuration
          config = read_configuration(filepath, True, ignore_option_errors, dist)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/config/pyprojecttoml.py", line 139, in read_configuration
          validate(subset, filepath)
        File "/tmp/pip-build-env-iekkrxzx/overlay/lib/python3.12/site-packages/setuptools/config/pyprojecttoml.py", line 60, in validate
          raise ValueError(f"{error}\n{summary}") from None
      ValueError: invalid pyproject.toml config: `project.optional-dependencies.qa[0]`.
      configuration error: `project.optional-dependencies.qa[0]` must be pep508
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

</details>

xylar replied:
Try `==` instead of `=`

xylar replied:
If that doesn't work comment out qa for now.

altheaden replied:
@forsyth2 I think it might be mad that the black version is being compared with a single equals sign instead of a double equals. We don't have a comparison like that in mache so I haven't needed to format it for the toml file yet, but that makes sense. Could you change the last two lists to these and see how it works?
```toml
qa = [
    "black==23.9.1",
    "flake8==6.1.0",
    "flake8-isort==6.1.0",
    "isort==5.12.0",
    "mypy==1.5.1",
    "pre-commit >=3.0.0",
    "types-PyYAML >=6.0.0",
]

dev = [
    "tbump==6.9.0",
    "ipykernel",
]
```

forsyth2 commented on pyproject.toml:
```diff
@@ -0,0 +1,90 @@
+]
+
+dependencies = [
+    "matplotlib-base >=3.8.2,<3.10",
```
This toml update seems to be causing an issue.
```
$ pip install .
Processing /gpfs/fs1/home/ac.forsyth2/ez/zppy-interfaces
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of zppy-interfaces to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement matplotlib-base<3.10,>=3.8.2 (from zppy-interfaces) (from versions: none)
ERROR: No matching distribution found for matplotlib-base<3.10,>=3.8.2
```

forsyth2 replied:
@altheaden Do you have an idea of what's going on here? No matter how I edit this line I can't seem to find something that `pip` will take.

forsyth2 replied:
It then appears to give a similar error about `xcdat` if I just remove `matplotlib-base` entirely.

xylar replied:
@forsyth2 Sorry. You should change this to just matplotlib without core here. That package doesn't exist on pypi, just on conda-forge.

xylar replied:
But do leave matplotlib-core in the dev.yml

xylar replied:
```suggestion
    "matplotlib >=3.8.2,<3.10",
```

forsyth2 commented on zppy_interfaces/global_time_series/__main__.py:
```diff
@@ -0,0 +1,117 @@
+
+
+# TODO: replace command line arguments with _get_cfg_parameters, like https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/parser/core_parser.py#L809
+def _get_args() -> Parameters:
```
The `zppy` side doesn't seem to be passing in arguments correctly.

The bash template, when filled out, shows:
```
zi-global-time-series --use_ocn True --global_ts_dir /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_\
original_8_output/test-642-working-env-20241105/v3.LR.historical_0051/post/scripts/global_time_series_1985-1995_dir\
 --input /lcrc/group/e3sm2/ac.wlin/E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeS\
eries_1985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-\
642-working-env-20241105/v3.LR.historical_0051 --experiment_name v3.LR.historical_0051 --figstr v3.LR.historical_00\
51 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,\
net_atm_energy_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --atmosphere_only False --plot\
s_atm None --plots_ice None --plots_lnd None --plots_ocn None --regions glb,n,s --results_dir ${results_dir} --star\
t_yr 1985 --end_yr 1995
```
But `/lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241105/v3.LR.historical_0051/post/scripts/global_time_series_1985-1995.o622622` shows:
```
zi-global-time-series: error: the following arguments are required: end_yr
```

If I try to enter `end_yr` manually on the `zppy_interfaces` side, it then raises an error about `start_yr`. So clearly the arguments aren't getting passed in properly.

xylar replied:
It looks like you have a lot of `\` line dividers in the middle of expressions.  I think that is going to cause giant problems.  You can only have ` \` between flags, not in the middle of them.

forsyth2 commented on zppy_interfaces/global_time_series/coupled_global.py:
```diff
@@ -0,0 +1,1044 @@
+        self.directory: str = directory
+
+        # `directory` will be of the form `{case_dir}/post/<component>/glb/ts/monthly/{ts_num_years_str}yr/`
+        # TODO: run this print statement here and in zppy main, and compare.
+        # TODO: double check we have the same xcdat version in the 2 cases.
+        # TODO: can we open these 3 folders from a tiny script?
+        # TODO: put in file names manually rather than using `*`?
+        print(f"error is here: {directory}*.nc")
```
I still need to debug why the ocean plots are getting stuck here but not in pre-split `zppy`.

forsyth2 replied:
```
$ conda activate zppy-interfaces-dev-20241104
$ cd /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_setup_only_output/test-642-2024-1105/v3.LR.historical_0051/post/ocn/glb/ts/monthly/5yr
$ ls
mocTimeSeries_1985-1995.nc  mpaso.glb.198501-198912.nc  mpaso.glb.199001-199412.nc
$ python
>>> import xcdat
>>> xcdat.open_mfdataset("*.nc", center_times=True)
# ValueError: Could not find any dimension coordinates to use to order the datasets for concatenation
```

```
$ source /lcrc/soft/climate/e3sm-unified/load_latest_e3sm_unified_chrysalis.sh
$ cd /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-main-20241105/v3.LR.historical_0051/post/ocn/glb/ts/monthly/5yr/
$ ls
mocTimeSeries_1985-1995.nc  mpaso.glb.198501-198912.nc  mpaso.glb.199001-199412.nc
$ python
>>> import xcdat
>>> xcdat.open_mfdataset("*.nc", center_times=True)
# Successful
```

forsyth2 replied:
Respectively:
```
>>> xcdat.__version__
'0.6.0'
```
```
>>> xcdat.__version__
'0.7.0'
```
Looks like it's the `xcdat` version.

forsyth2 replied:
Since we're setting `"xcdat >=0.7.0,<1.0",`, this problem should get resolved.

tomvothecoder replied:
FYI I will be releasing xCDAT v0.7.3 soon which sets `numpy >=2.0.0`. I will let you know soon so you can set that as the min version here.

xylar replied:
Ah, yes, the specs here are probably currently unsovlable.  @forsyth2, you probably got xcdat 0.6.0 before because that's before @tomvothecoder put an upper bound on numpy, so it can technically be installed with numpy 2.0 even though it won't actually work.  (We could patch things on conda-forge to fix that but probably not worth it.)

forsyth2 reviewed (commented):
Ok I think we're close to being able to replicate existing functionality using the split-repo approach.

3 issues remaining, noted as part of this review.

tomvothecoder commented on pyproject.toml:
```diff
@@ -0,0 +1,90 @@
+    "netcdf4",
+    "numpy >=2.0,<3.0",
+    "xarray >=2023.02.0",
+    "xcdat >=0.7.0,<1.0",
```
```suggestion
    "xcdat >=0.7.3,<1.0",
```

xcdat v0.7.3 is now available on `conda-forge` with the same numpy 2.0 constraint: https://anaconda.org/conda-forge/xcdat

tomvothecoder commented on conda/dev.yml:
```diff
@@ -0,0 +1,43 @@
+  - netcdf4
+  - numpy >=2.0,<3.0
+  - xarray >=2023.02.0
+  - xcdat >=0.7.0,<1.0
```
```suggestion
  - xcdat >=0.7.3,<1.0
```

xcdat v0.7.3 is now available on `conda-forge` with the same numpy 2.0 constraint: https://anaconda.org/conda-forge/xcdat

xylar commented:
@forsyth2, haven't heard anything further on this.  Has this just had to move to the back burner or are you stuck anywhere that @altheaden, @tomvothecoder or I could help with?

forsyth2 commented:
Ok, I'm getting back into this after a full two weeks away from it.

### Checking where we were at

I addressed comments on the build files (see latest commit)

I created a new environment to test in:
```bash
cd ~/ez/zppy-interfaces
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_interfaces_dev_pr_1_20241120
conda activate zppy_interfaces_dev_pr_1_20241120
pip install .
```

While running `conda env create`, I got 
```
==> WARNING: A newer version of conda exists. <==
  current version: 4.12.0
  latest version: 24.9.2

Please update conda by running

    $ conda update -n base conda
```
which I assume is because I had previously been using `mamba`, so I ran `conda update -n base conda`.

I ran the unit tests for `zppy-interfaces`:
```
$ python -u -m unittest tests/global_time_series/test_global_time_series.py
.......
----------------------------------------------------------------------
Ran 7 tests in 0.001s

OK
```

I ran the `if __name__ == "__main__":` placeholder for actually being launched by `zppy`:
```bash
cd zppy_interfaces/global_time_series
python __main__.py
```
That gives:
```
Update time series figures
error is here: /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_setup_only_output/test-642-2024-1105/v3.LR.historical_0051/post/atm/glb/ts/monthly/5yr/*.nc
Exception ignored in: <function TS.__del__ at 0x155118e57560>
Traceback (most recent call last):
  File "/home/ac.forsyth2/miniconda3/envs/zppy_interfaces_dev_pr_1_20241120/lib/python3.12/site-packages/zppy_interfaces/global_time_series/coupled_global.py", line 125, in __del__
    self.f.close()
    ^^^^^^
AttributeError: 'TS' object has no attribute 'f'
```

This is what the comment `# TODO: fix readTS errors` is referring to in `zppy_interfaces/global_time_series/__main__.py`.

Now, for integrating with `zppy`. I also attempted to create a new environment here:
```bash
cd ~/ez/zppy
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_dev_pr_642_20241120v2 
# v2 because I originally tried this before updating conda, and it wasn't done after 30 minutes. 
# It appears it's still taking quite a while though. The only reason I can think of here is using conda rather than mamba.
```

### What's left to do?

I'm hoping to have time to address the following points tomorrow and Friday.

Action items from above:
- [ ] Resolve `readTS` error
- [ ] Actually get a new zppy dev environment to build and then do the following:

```bash
cd ~/ez/zppy
conda activate zppy_dev_pr_642_20241120
pip install .
```
Then:
```bash
# Update `UNIQUE_ID` in `zppy/tests/integration/utils.py`
python tests/integration/utils.py
zppy -c tests/integration/generated/test_min_case_global_time_series_original_8_chrysalis.cfg
```
It looks like parameters weren't getting passed in correctly though: https://github.com/E3SM-Project/zppy-interfaces/pull/1/files#r1830208057

- [ ] Update code to use a config file rather than passing in parameters as a giant command.

Remaining action items from https://github.com/E3SM-Project/zppy-interfaces/pull/1#pullrequestreview-2414462783:
- [ ] Run the `global_time_series` task in zppy (i.e., make sure `zppy` can correctly call `zppy-interfaces`).
- [ ] Create a `tests/` directory and add some integration tests. `zppy` should simply test that it can invoke `zppy-interfaces global-time-series`. `zppy-interfaces` however should test that Global Time Series plots actually generate correctly. As for unit tests, https://github.com/E3SM-Project/zppy/pull/616 will do that once it's ported to this repo.
- [ ] Add documentation for this repo/package.

After these PRs merge:
- [ ] Adjust the land support/Viewer generation PR (https://github.com/E3SM-Project/zppy/pull/616) accordingly
- [ ] Help integrate the PMP PR (https://github.com/E3SM-Project/zppy/pull/640) accordingly

xylar commented:
@forsyth2, it sounds like you should delete whatever installation of mambaforge/miniconda/whatever that you are currently using and start fresh.  Download miniforge3:
https://github.com/conda-forge/miniforge?tab=readme-ov-file#miniforge3
Install that in ~/miniforge3 (unless you have a reason to install it somewhere else) and use that from now on.  Please update documentation everywhere to refer to miniforge3 rather than either miniconda or mambaforge.  Newer versions of conda, like that from miniforge3, will provide you with libmamba as the default package solver, which means there is no added value in using the old (deprecated) mamba command.  Solving your environment should be fast.

xylar commented:
> ```
> cd zppy_interfaces/global_time_series
> python __main__.py
> ```

While that's fine, it should also be possible to use the entry point after you've run `pip install .`:
```
zi-global-time-series ...
```

forsyth2 commented:
## Updating conda build

Previously had this defined in `.bashrc` to set up conda:
```
lcrc_conda()
{
  # >>> conda initialize >>>                                                                                                                                                     
  # !! Contents within this block are managed by 'conda init' !!                                                                                                                 
  __conda_setup="$('/home/ac.forsyth2/miniconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
  if [ $? -eq 0 ]; then
      eval "$__conda_setup"
  else
      if [ -f "/home/ac.forsyth2/miniconda3/etc/profile.d/conda.sh" ]; then
          # . "/home/ac.forsyth2/miniconda3/etc/profile.d/conda.sh"  # commented out by conda initialize                                                                         
          echo ""
      else
          export PATH="/home/ac.forsyth2/miniconda3/bin:$PATH"
      fi
  fi
  unset __conda_setup
  # <<< conda initialize <<<                                                                                                                                                     
}
```

Deleted that block. Ran `source ~/.bashrc`

Went to https://github.com/conda-forge/miniforge (full link was giving a security error)

```
wget "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
bash Miniforge3-$(uname)-$(uname -m).sh
```

```
Miniforge3 will now be installed into this location:
/home/ac.forsyth2/miniforge3
```

```
To activate this environment, use:

    micromamba activate /gpfs/fs1/home/ac.forsyth2/miniforge3

Or to execute a single command in this environment, use:

    micromamba run -p /gpfs/fs1/home/ac.forsyth2/miniforge3 mycommand

installation finished.
Do you wish to update your shell profile to automatically initialize conda?
This will activate conda on startup and change the command prompt when activated.
If you'd prefer that conda's base environment not be activated on startup,
   run the following command when conda is activated:

conda config --set auto_activate_base false

You can undo this by running `conda init --reverse $SHELL`? [yes|no]
```

```
You have chosen to not have conda modify your shell scripts at all.
To activate conda's base environment in your current shell session:

eval "$(/home/ac.forsyth2/miniforge3/bin/conda shell.YOUR_SHELL_NAME hook)" 

To install conda's shell functions for easier access, first activate, then:

conda init

Thank you for installing Miniforge3!
```

```
eval "$(/home/ac.forsyth2/miniforge3/bin/conda shell.bash hook)"
conda init
# ==> For changes to take effect, close and re-open your current shell. <==
```

Now, this is at the end of my `.bashrc`:
```
# >>> conda initialize >>>                                                                                                                                                       
# !! Contents within this block are managed by 'conda init' !!                                                                                                                   
__conda_setup="$('/gpfs/fs1/home/ac.forsyth2/miniforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh" ]; then
        . "/gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh"
    else
        export PATH="/gpfs/fs1/home/ac.forsyth2/miniforge3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<  
```

Wrapped that in a function to call myself rather than having it auto-run:
```
lcrc_conda()
{
  # >>> conda initialize >>>                                                                                                                                                       
  # !! Contents within this block are managed by 'conda init' !!                                                                                                                   
  __conda_setup="$('/gpfs/fs1/home/ac.forsyth2/miniforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
  if [ $? -eq 0 ]; then
      eval "$__conda_setup"
  else
      if [ -f "/gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh" ]; then
          . "/gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh"
      else
          export PATH="/gpfs/fs1/home/ac.forsyth2/miniforge3/bin:$PATH"
      fi
  fi
  unset __conda_setup
  # <<< conda initialize <<<  
}
```

## Creating new environments
```
cd ~/ez/zppy-interfaces
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_interfaces_dev_pr_1_20241121
conda activate zppy_interfaces_dev_pr_1_20241121
pip install .

cd ~/ez/zppy
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_dev_pr_642_20241121
conda activate zppy_dev_pr_642_20241121
pip install .
```

## Testing/debugging

```
$ zi-global-time-series
usage: zi-global-time-series <args>
zi-global-time-series: error: the following arguments are required: use_ocn, input, input_subdir, moc_file, case_dir, experiment_name, figstr, color, ts_num_years, plots_original, atmosphere_only, plots_atm, plots_ice, plots_lnd, plots_ocn, nrows, ncols, results_dir, regions, start_yr, end_yr
```
Re: https://github.com/E3SM-Project/zppy-interfaces/pull/1#issuecomment-2490441424 -- running the entry point requires the parameters be passed in on the command line. (It runs the `main` function, not the `if __name__ == "__main__":` block).

Edit `zppy/tests/integration/utils.py`:
```
UNIQUE_ID = "test-642-working-env-20241121"
```

```bash
python tests/integration/utils.py
zppy -c tests/integration/generated/test_min_case_global_time_series_original_8_chrysalis.cfg
# Wait for that to finish
cd /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241121/v3.LR.historical_0051/post/scripts
cat global_time_series_1985-1995.o632171 
```
That gives:
```
usage: zi-global-time-series <args>
zi-global-time-series: error: the following arguments are required: end_yr
Copy images to directory
cp: cannot stat '*.pdf': No such file or directory
cp: cannot stat '*.png': No such file or directory
```
In the `bash` file:
```
zi-global-time-series --use_ocn True --global_ts_dir /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241121/v3.LR.histori\
cal_0051/post/scripts/global_time_series_1985-1995_dir --input /lcrc/group/e3sm2/ac.wlin/E3SMv3/v3.LR.historical_0051 --input_subdir archive/ocn/hist --moc_file mocTimeSeries_1\
985-1995.nc --case_dir /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241121/v3.LR.historical_0051 --experiment_name v3.\
LR.historical_0051 --figstr v3.LR.historical_0051 --color Blue --ts_num_years 5 --plots_original net_toa_flux_restom,global_surface_air_temperature,toa_radiation,net_atm_energy\
_imbalance,change_ohc,max_moc,change_sea_level,net_atm_water_imbalance --atmosphere_only False --plots_atm None --plots_ice None --plots_lnd None --plots_ocn None --regions glb\
,n,s --results_dir ${results_dir} --start_yr 1985 --end_yr 1995
```
Re: https://github.com/E3SM-Project/zppy-interfaces/pull/1#discussion_r1830397828 -- the `\` are a product of my terminal size and are not actually present in the command.


Made some edits in the `zppy` repo.

```bash
cd /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241121/v3.LR.historical_0051/post/scripts
rm -rf global_time_series_1985-1995* # Allow us to re-run just the global_time_series part
cd /home/ac.forsyth2/ez/zppy
pip install .
python tests/integration/utils.py
zppy -c tests/integration/generated/test_min_case_global_time_series_original_8_chrysalis.cfg
```

The `.o` file still gives:
```
zi-global-time-series: error: the following arguments are required: start_yr, end_yr
```

Continued various debugging attempts on both the `zppy` and `zppy-interfaces` sides. Eventually got:
```
zi-global-time-series: error: unrecognized arguments: --use_ocn --input --input_subdir --moc_file --case_dir --experiment_name --figstr --color --ts_num_years --plots_original --atmosphere_only --plots_atm --plots_ice --\
plots_lnd --plots_ocn --nrows --ncols --results_dir --regions --start_yr --end_yr
```
which seems to be the opposite problem.

Looks like the issue is the args actually have to be defined with `--` for the arg parser to work.

## zppy successfully generating plots by calling zppy-interfaces
Actually got zppy to call zppy-interfaces successfully. Still have the TS error on the zppy-interfaces side though.

After more debugging, was successful in getting plots to show up at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_original_8_www/test-642-working-env-20241121/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/

forsyth2 commented:
## Cleaning up the code

Made edits to both `zppy` and `zppy-interfaces` (Commit for each is named "Updates").

This included build updates for `zppy-interfaces`, so let's create a new env:
```
cd ~/ez/zppy-interfaces
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_interfaces_dev_pr_1_20241121v2
conda activate zppy_interfaces_dev_pr_1_20241121v2
pip install .
```

## Rerun min-case
Rerun the min-case to make sure everything still works:

Edit `zppy/tests/integration/utils.py`:
```
UNIQUE_ID = "test-642-working-env-20241121v2"
```
```
        "global_time_series_environment_commands": "source /gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh; conda activate zppy_interfaces_dev_pr_1_20241121v2",
```
```
python tests/integration/utils.py
pip install .
zppy -c tests/integration/generated/test_min_case_global_time_series_original_8_chrysalis.cfg
# Wait for that to finish
cd /lcrc/group/e3sm/ac.forsyth2/zppy_min_case_global_time_series_original_8_output/test-642-working-env-20241121v2/v3.LR.historical_0051/post/scripts
grep -v "OK" *status
```
Results at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_original_8_www/test-642-working-env-20241121v2/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/

## Run unit tests
```
cd ~/ez/zppy-interfaces
python -u -m unittest tests/global_time_series/test_global_time_series.py
```

```
cd ~/ez/zppy
python -u -m unittest tests/test_*.py
```
`test_sections` and `test_subsections` had to be updated.

All unit tests are now passing.

forsyth2 reviewed (commented):
I think I have the initial implementation of the `zppy`/`zppy-interfaces` refactor done.

[Required reviewers] @xylar @chengzhuzhang I'd like to get approval from both of you before merging this.

[Optional reviewers] @tomvothecoder @altheaden you may want to review the conda/build portions of this PR.

## Remaining to-do

After the refactor PRs merge:
- [ ] Run `zppy`'s Weekly tests on `main` to make sure everything is still working properly after the merges.
- [ ] Adjust the land support/Viewer generation PR (https://github.com/E3SM-Project/zppy/pull/616) accordingly
- [ ] Help integrate the PMP PR (https://github.com/E3SM-Project/zppy/pull/640) accordingly

Add these as GitHub issues to get to later:
- [ ] Add documentation, according to https://docs.e3sm.org/zppy/_build/html/main/contributing.html#initial-setup-obsolete-for-reference-only. (Looks like it's better to do that as its own PR) [definitely need to do this before the Unified release!]
- [ ] Add option to specify parameters through cfg rather than command line. E.g., https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/parser/core_parser.py#L809
- [ ] Switch pre-commit checks to `ruff`, as in https://github.com/xCDAT/xcdat/pull/702/files.
- [ ] Rethink how testing is done in `zppy` & `zppy-interfaces`. `zppy` should simply test that it can invoke `zppy-interfaces global-time-series`. `zppy-interfaces` however should test that Global Time Series plots actually generate correctly. Maybe have `zppy` test that `zppy` is passing in the right parameters and then have `zppy-interfaces` actually do the image comparisons of the plots, using pre-generated `ts` and `mpas_analysis` output.

## Helpful details for reviewers
This is actually 2 PRs. This one and https://github.com/E3SM-Project/zppy/pull/642.

### zppy
On https://github.com/E3SM-Project/zppy/pull/642:

- Anything under `tests/integration/generated` can be ignored. Those are just auto-generated testing files.
- I added a `global_time_series`-specific `environment_commands` to the test suite, so that `zppy-interfaces` could be used.
  - This is reflected in the 6 template `cfg`s (3 of these templates are "min-case" files which I use for debugging, and the other 3 are the big tests we run roughly weekly on `main`. For this PR, I've been debugging with the `min_case_global_time_series_original_8` min-case.)
  - Also reflected in `tests/integration/utils.py`(This file is used to make run-able `cfg`s out of the templates).
- I organized `zppy/templates` to be more logical.
  - There is now a subdirectory `zppy/templates/inclusions` which is for adding text word-for-word. 
  - `default.ini` and the campaign files are now kept in `zppy/defaults`.
  - `zppy/examples` includes files that aren't actually used anywhere and thus serve only as examples.
  - The `global_time_series`-specific files (`coupled_global.py`, `ocean_month.py`, `readTS.py`) have been moved over to `zppy-interfaces/zppy_interfaces/global_time_series`, out of the `zppy` repo completely. Note, the `readTS.py` file has been included into `coupled_global.py` and other utilities have been pulled out into a new `utils.py`.
  - The `jinja2` template bash scripts remain in `zppy/templates` itself. 
  - These directory changes can be seen throughout the PR, including in the unit test changes.
- Much of `zppy/templates/global_time_series.bash` has been moved over to `zppy_interfaces/global_time_series/__main__.py` as `python` code rather than `bash` code.

### zppy-interfaces
This PR

- Conda/build updates: `.flake8`, `.pre-commit-config.yaml`, `conda/dev.yml`, `pyproject.toml`, `zppy_interfaces/version.py`
- I noticed many changes in https://github.com/E3SM-Project/zppy/pull/616 were actually more general and not specifically about Land support and/or Viewer generation. I've pulled out those changes (notably refactoring of `coupled_global.py` into more concise, composable functions & testing those functions in unit tests)
  - Unit test: `tests/global_time_series/test_global_time_series.py`
  - `zppy_interfaces/global_time_series/coupled_global.py` and `zppy_interfaces/global_time_series/utils.py` reflect these changes.
- The only entry point right now is `zi-global-time-series` which runs `zppy_interfaces/global_time_series/__main__.py main`. 
  - Please note that this modifies the case directory (only the `posts/` subdirectory though), which could be seen as modifying the input. However, this was the case before the refactor as well. Pre-refactor, the case dir was defined to be `{{ output }}`, so we knew we were modifying the directory the user specified as output (which they could have set to be identical to `{{ input }}`). Now, we sort of assume `output dir = input dir = case dir`, relying on `zppy` to pass in the right output dir.
  - The `if __name__ == "__main__":` block of that file is for debugging manually, and is not called by `zppy`.

tomvothecoder commented on conda/dev.yml:
```diff
@@ -0,0 +1,43 @@
+  # =======================
+  # Run `pre-commit autoupdate` to get the latest pinned versions of 'rev' in
+  # `.pre-commit.config.yaml`, then update the pinned versions here.
+  - black=23.9.1
+  - flake8=6.1.0
+  - flake8-isort=6.1.0
+  - isort=5.12.0
+  - mypy=1.5.1
+  - pre-commit >=3.0.0
+  - types-PyYAML >=6.0.0
```
```suggestion
  - black=24.10.0
  - flake8=7.1.1
  - flake8-isort=6.1.0
  - isort=5.13.2
  - mypy=1.11.2
  - pre-commit >=3.0.0
  - types-PyYAML >=6.0.0
```

Versions should align with `pre-commit-config.yaml`

tomvothecoder commented on pyproject.toml:
```diff
@@ -0,0 +1,123 @@
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13",
+
```
```suggestion
```

tomvothecoder commented on pyproject.toml:
```diff
@@ -0,0 +1,123 @@
+    "sphinx-multiversion",
+]
+
+qa = [
+    "black==23.9.1",
+    "flake8==6.1.0",
+    "flake8-isort==6.1.0",
+    "isort==5.12.0",
+    "mypy==1.5.1",
+    "pre-commit >=3.0.0",
+    "types-PyYAML >=6.0.0",
+]
```
```suggestion
qa = [
    "black==24.10.1",
    "flake8==7.1.1",
    "flake8-isort==6.1.0",
    "isort==5.13.2",
    "mypy==1.11.2",
    "pre-commit >=3.0.0",
    "types-PyYAML >=6.0.0",
]
```

Versions should align with `.pre-commit-config.yaml`.

tomvothecoder commented on pyproject.toml:
```diff
@@ -0,0 +1,123 @@
+
+[tool.black]
+line-length = 88
+target-version = ['py36']
```
```suggestion
target-version = ['py39']
```

This should be the minimum version of Python (3.9 for `zppy` currently).

tomvothecoder commented on tests/global_time_series/test_global_time_series.py:
```diff
@@ -0,0 +1,288 @@
+import unittest
```
I prefer using `pytest` over the standard lib `unittest` module. Seems to make testing less verbose, quicker to read, and easier to write for me. You can probably use GH Copilot to convert these tests into `pytest` (up to you).

tomvothecoder commented on zppy_interfaces/global_time_series/coupled_global.py:
```diff
@@ -0,0 +1,1036 @@
+        try:
+            ts_object: TS = TS(exp[exp_key])
+        except Exception as e:
+            print(e)
+            print(f"TS object could not be created for {exp_key}={exp[exp_key]}")
```
Instead of using print statements, you should use the Python `logger` module for easier debugging.

tomvothecoder reviewed (commented):
My brief review of the DevOps related files with some comments and suggestions. 

I recommend setting up a GH Actions build workflow similar to `zppy` that runs the `pre-commit` hooks and unit tests: https://github.com/E3SM-Project/zppy/blob/main/.github/workflows/build_workflow.yml

forsyth2 commented:
Thanks @tomvothecoder! I've implemented your suggestions on both PRs.

<details> 

<summary> Details of testing </summary>

```
cd ~/ez/zppy-interfaces
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_interfaces_dev_pr_1_20241121v4
conda activate zppy_interfaces_dev_pr_1_20241121v4
pip install .
pytest tests/global_time_series

cd ~/ez/zppy
conda clean --all --y
conda env create -f conda/dev.yml -n zppy_dev_pr_642_20241121v4
conda activate zppy_dev_pr_642_20241121v4
pip install .
pytest tests/test*.py # Skip integration test directory
```
That gives a warning -- one we're already escaping -- but all tests pass
```
===================================================================================================== warnings summary ======================================================================================================
zppy/utils.py:191
  /gpfs/fs1/home/ac.forsyth2/ez/zppy/zppy/utils.py:191: DeprecationWarning: invalid escape sequence \.
    tmp = re.sub("\.[^.]*\.nc$", "", tmp)  # noqa: W605

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================== 23 passed, 1 warning in 0.20s ==============================================================================================
```

```
# Now, run min case
# Edit tests/integration/utils.py
# UNIQUE_ID = "test-642-working-env-20241121v4"
# "global_time_series_environment_commands": "source /gpfs/fs1/home/ac.forsyth2/miniforge3/etc/profile.d/conda.sh; conda activate zppy_interfaces_dev_pr_1_20241121v4",
python tests/integration/utils.py
zppy -c tests/integration/generated/test_min_case_global_time_series_original_8_chrysalis.cfg
```

Plots are generated at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_original_8_www/test-642-working-env-20241121v4/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/.

After some edits to the `build_workflow.yml` files I was able to get checks passing on https://github.com/E3SM-Project/zppy/pull/642 and on this PR.

</details>

xylar reviewed (approved):
@forsyth2, this looks great!  On small suggested change.

forsyth2 commented:
Thanks for the reviews @xylar!

chengzhuzhang reviewed (approved):
Hi @forsyth2 this is exciting change. Thanks for making it happening. I think @xylar and @tomvothecoder had solid reviews for DevOp codes. and things looks promising. I'm approving this PR, it addresses the initial implementation. 

Regarding to usability, I'm wondering if it is general enough, for instance, does `zi-global-time-series` support land only simulation with easy syntax? Maybe the `coupled_global.py` can be polished a bit. Anyway, these can be addressed in a future  PR.

forsyth2 commented:
Thanks @chengzhuzhang!

> Regarding to usability, I'm wondering if it is general enough, for instance, does `zi-global-time-series` support land only simulation with easy syntax? Maybe the `coupled_global.py` can be polished a bit. Anyway, these can be addressed in a future PR. 

- https://github.com/E3SM-Project/zppy/pull/616 will take care of adding Land support.
- The issue (which I still need to create) for using a cfg rather than command line options will go a long way towards generalized usability.

chengzhuzhang commented:
> * The issue (which I still need to create) for using a cfg rather than command line options will go a long way towards generalized usability.

Thank you @forsyth2 , command line options are good to have as well, just we need to think through  all the options and make them intuitive to use.

forsyth2 commented:
I added the issues mentioned in https://github.com/E3SM-Project/zppy-interfaces/pull/1#pullrequestreview-2452835664: #2, #3, #4, #5

forsyth2 commented:
It looks like any remaining comments/concerns can be addressed in later pull requests, so I think it's safe to say we have the initial implementation done. I'm going to merge this PR and https://github.com/E3SM-Project/zppy/pull/642. 

The plan after that is:
- [ ] Run `zppy`'s Weekly tests on `main` to make sure everything is still working properly after the merges.
- [ ] Adjust the land support/Viewer generation PR (https://github.com/E3SM-Project/zppy/pull/616) accordingly
- [ ] Help integrate the PMP PR (https://github.com/E3SM-Project/zppy/pull/640) accordingly -- I have a meeting scheduled to discuss this.
- [ ] The new `zppy-interfaces` issues I mentioned above.
